File: go/test/plugin_listing_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestPluginListingEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.PluginListing(nil)
		if ent == nil {
			t.Fatal("expected non-nil PluginListingEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := plugin_listingBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "plugin_listing." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_PLUGIN_LISTING_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		pluginListingRef01Ent := client.PluginListing(nil)
		pluginListingRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "plugin_listing"}, setup.data), "plugin_listing_ref01"))
		pluginListingRef01Data["id_plugin"] = setup.idmap["id_plugin01"]

		pluginListingRef01DataResult, err := pluginListingRef01Ent.Create(pluginListingRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		pluginListingRef01Data = core.ToMapAny(entityData(pluginListingRef01DataResult))
		if pluginListingRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if pluginListingRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		pluginListingRef01DataUp0Up := map[string]any{
			"id": pluginListingRef01Data["id"],
			"id_plugin": setup.idmap["id_plugin"],
		}

		pluginListingRef01MarkdefUp0Name := "description"
		pluginListingRef01MarkdefUp0Value := fmt.Sprintf("Mark01-plugin_listing_ref01_%d", setup.now)
		pluginListingRef01DataUp0Up[pluginListingRef01MarkdefUp0Name] = pluginListingRef01MarkdefUp0Value

		pluginListingRef01ResdataUp0Result, err := pluginListingRef01Ent.Update(pluginListingRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		pluginListingRef01ResdataUp0 := core.ToMapAny(entityData(pluginListingRef01ResdataUp0Result))
		if pluginListingRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if pluginListingRef01ResdataUp0["id"] != pluginListingRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if pluginListingRef01ResdataUp0[pluginListingRef01MarkdefUp0Name] != pluginListingRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", pluginListingRef01MarkdefUp0Name, pluginListingRef01ResdataUp0[pluginListingRef01MarkdefUp0Name])
		}

	})
}

func plugin_listingBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "plugin_listing", "PluginListingTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read plugin_listing test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse plugin_listing test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"plugin_listing01", "plugin_listing02", "plugin_listing03", "plugin01", "plugin02", "plugin03", "id_plugin01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_PLUGIN_LISTING_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_PLUGIN_LISTING_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_PLUGIN_LISTING_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add id_plugin alias for update test.
	if idmapResolved["id_plugin"] == nil {
		idmapResolved["id_plugin"] = idmapResolved["id_plugin01"]
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/plugin_listing_entity_test.lua
-- PluginListing entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("trello_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("PluginListingEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:PluginListing(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = plugin_listing_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "plugin_listing." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_PLUGIN_LISTING_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local plugin_listing_ref01_ent = client:PluginListing(nil)
    local plugin_listing_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.plugin_listing"), "plugin_listing_ref01"))
    plugin_listing_ref01_data["id_plugin"] = setup.idmap["id_plugin01"]

    local plugin_listing_ref01_data_result, err = plugin_listing_ref01_ent:create(plugin_listing_ref01_data, nil)
    assert.is_nil(err)
    plugin_listing_ref01_data = helpers.to_map(type(plugin_listing_ref01_data_result) == 'table' and plugin_listing_ref01_data_result.data_get and plugin_listing_ref01_data_result:data_get() or plugin_listing_ref01_data_result)
    assert.is_not_nil(plugin_listing_ref01_data)
    assert.is_not_nil(plugin_listing_ref01_data["id"])

    -- UPDATE
    local plugin_listing_ref01_data_up0_up = {
      id = plugin_listing_ref01_data["id"],
      ["id_plugin"] = setup.idmap["id_plugin"],
    }

    local plugin_listing_ref01_markdef_up0_name = "description"
    local plugin_listing_ref01_markdef_up0_value = "Mark01-plugin_listing_ref01_" .. tostring(setup.now)
    plugin_listing_ref01_data_up0_up[plugin_listing_ref01_markdef_up0_name] = plugin_listing_ref01_markdef_up0_value

    local plugin_listing_ref01_resdata_up0_result, err = plugin_listing_ref01_ent:update(plugin_listing_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local plugin_listing_ref01_resdata_up0 = helpers.to_map(type(plugin_listing_ref01_resdata_up0_result) == 'table' and plugin_listing_ref01_resdata_up0_result.data_get and plugin_listing_ref01_resdata_up0_result:data_get() or plugin_listing_ref01_resdata_up0_result)
    assert.is_not_nil(plugin_listing_ref01_resdata_up0)
    assert.are.equal(plugin_listing_ref01_resdata_up0["id"], plugin_listing_ref01_data_up0_up["id"])
    assert.are.equal(plugin_listing_ref01_resdata_up0[plugin_listing_ref01_markdef_up0_name], plugin_listing_ref01_markdef_up0_value)

  end)
end)

function plugin_listing_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/plugin_listing/PluginListingTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read plugin_listing test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "plugin_listing01", "plugin_listing02", "plugin_listing03", "plugin01", "plugin02", "plugin03", "id_plugin01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("TRELLO_TEST_PLUGIN_LISTING_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["TRELLO_TEST_PLUGIN_LISTING_ENTID"] = idmap,
    ["TRELLO_TEST_LIVE"] = "FALSE",
    ["TRELLO_TEST_EXPLAIN"] = "FALSE",
    ["TRELLO_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["TRELLO_TEST_PLUGIN_LISTING_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["id_plugin"] == nil then
    idmap_resolved["id_plugin"] = idmap_resolved["id_plugin01"]
  end

  if env["TRELLO_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["TRELLO_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["TRELLO_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["TRELLO_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
